test/exe/win32_scm_test.cc

```diff
@@ -0,0 +1,169 @@
+  stopService();
+  auto log_path = TestEnvironment::temporaryDirectory() + "/output.txt";
+  auto commandLine =
+      fmt::format("{} --config-path {} -l warn --log-path {}", envoy_exe_, config_file_, log_path);
+  changeServicePath(commandLine);
+  startService();
+  stopService();
+  remove(log_path.c_str());
+}
+
+TEST_P(ScmTest, EnvoySetsInvalidArgError) {
+  stopService();
+  auto commandLine = fmt::format("{} this is an invalid envoy command line", envoy_exe_);
+  changeServicePath(commandLine);
+  EXPECT_TRUE(StartServiceA(service_handle_, NULL, NULL))
+      << fmt::format("Could not start service, error {}", ::GetLastError());
+  waitForServiceStatus(SERVICE_STOPPED);
+  SERVICE_STATUS_PROCESS service_status;
+  DWORD bytes_needed;
+  EXPECT_TRUE(QueryServiceStatusEx(service_handle_, SC_STATUS_PROCESS_INFO, (LPBYTE)&service_status,
+                                   sizeof(SERVICE_STATUS_PROCESS), &bytes_needed))
+      << fmt::format("Could not query service status, error {}", ::GetLastError());
+  EXPECT_EQ(service_status.dwWin32ExitCode, ERROR_SERVICE_SPECIFIC_ERROR);
+  EXPECT_EQ(service_status.dwServiceSpecificExitCode, E_INVALIDARG);
+}
+
+TEST_P(ScmTest, EnvoySetsEnvoyExceptionError) {
+  stopService();
+  auto log_path = "invalid/path/";
+  auto commandLine =
+      fmt::format("{} --config-path {} -l warn --log-path {}", envoy_exe_, config_file_, log_path);
+  changeServicePath(commandLine);
+  EXPECT_TRUE(StartServiceA(service_handle_, NULL, NULL))
+      << fmt::format("Could not start service, error {}", ::GetLastError());
+  waitForServiceStatus(SERVICE_STOPPED);
+  SERVICE_STATUS_PROCESS service_status;
+  DWORD bytes_needed;
+  EXPECT_TRUE(QueryServiceStatusEx(service_handle_, SC_STATUS_PROCESS_INFO, (LPBYTE)&service_status,
+                                   sizeof(SERVICE_STATUS_PROCESS), &bytes_needed))
+      << fmt::format("Could not query service status, error {}", ::GetLastError());
+  EXPECT_EQ(service_status.dwWin32ExitCode, ERROR_SERVICE_SPECIFIC_ERROR);
+  EXPECT_EQ(service_status.dwServiceSpecificExitCode, E_FAIL);
+}
+
+TEST_P(ScmTest, EnvoyMergesArgumentsFromCreateAndStart) {
+  stopService();
+  auto log_path = TestEnvironment::temporaryDirectory() + "/output.txt";
+  remove(log_path.c_str());
+  auto commandLine = fmt::format("{} --config-path {} -l warn", envoy_exe_, config_file_);
+  changeServicePath(commandLine);
+  DWORD status = getServiceStatus();
+  if (status == SERVICE_RUNNING) {
+    stopService();
+  }
+
+  const char* extra_args[] = {"--log-path", log_path.c_str()};
+  EXPECT_TRUE(StartServiceA(service_handle_, 2, extra_args))
+      << fmt::format("Could not start service, error {}", ::GetLastError());
+  waitForServiceStatus(SERVICE_RUNNING);
+  std::ifstream genereated_file(log_path.c_str());
+  EXPECT_TRUE(genereated_file.good());
+  stopService();
+  genereated_file.close();
+  remove(log_path.c_str());
+}
+
+#endif
+
+} // namespace Envoy
```

Comment:
Watch for missing newlines at EOF (happens easily on win32).

Reply:
thanks good catch